packages/dds/tree/docs/wip/schema-evolution.md

```diff
@@ -0,0 +1,583 @@
+Allowing legacy code to be removed earlier improves codebase health, aids developer velocity and reduces the size of shipped code.
+
+To facilitate this, an application can write a schema adapter.
+The adapter takes a tree of data that conforms to one schema and transforms it to conform to another schema.
+In the context of a schema migration, this means that the application can write code (e.g. the code that binds the view schema to the UI) _only against the new view schema_.
+The code that binds to the old view schema is deleted at the same time that the new view schema and the new adapter are added.
+This is a net benefit because it's simpler to write the adapter's code than it is to write the UI binding code.
+During the migration, when the application loads a document with the old stored schema, it first adapts the data in the document to the new view schema before the data is read by the app.
+
+This replaces the view schema bifurcation with the adapter and changes the rollout sequence to:
+
+1. Initially, the documents are using stored schema A and the application is written against view schema A.
+2. Write an adapter from schema A to schema B, which will adapt the data in _stored_ schema A into the data in _view_ schema B.
+   Replace view schema A with view schema B and deploy a new version of the application.
+3. Wait for [client saturation](#migration), then enable the upgrading from stored schema A to stored schema B.
+4. If applicable, remove the adapter from the source code and redeploy after [document saturation](#migration).
+5. The migration is complete.
+
+```mermaid
+gantt
+dateFormat D
+axisFormat Step %e
+tickInterval 1day
+Maintain view schema A in source code                      : 1, 1d
+Maintain A -> B adapter in source code (read/write) : 2, 1d
+Maintain A -> B adapter in source code (read-only)  : 2, 2d
+Documents contain stored schema A                          : 1, 3d
+Maintain view schema B in source code                      : 2, 3d
+Documents contain stored schema B                          : 3, 2d
+```
+
+> Similar to a staged rollout without an adapter, read-only clients must preserve the adapter for more of the migration than read/write clients.
+  Therefore the advantages to using an adapter vs. not (i.e. maintaining the code bifurcation to support the old view schema) are extra valuable for applications with read-only clients.
+
+---
+
+An application could even write a "reverse" adapter to adapt the new schema to the old schema.
+This is useful if the application code wishes to move to a new view schema - and begin development on binding the UI to the model as soon as possible - but it can't do any upgrades to the stored schema of the documents yet.
+The application can use an adapter from schema A to schema B to read the document, and an adapter from schema B to schema A to write to the document.
```

Comment:
I got lost here. I assumed (and left the comment above) that was included automatically in writing an adapter - the bi-directionality of the adapter.
If it's not the case, then I think the steps / description in earlier section is incorrect, because app would need (in addition to having adapters and shipping their usage dark) also maintain code that can r/w in schema A directly. Only when client reaches 100% saturation (and makes a call that there is enough data that it's not going back) of clients doing conversion A -> B on open, only then clients can remove direct support of A and start using one-directional adapter.